### PR TITLE
dashboard: deep state update, footer

### PR DESCRIPTION
**Pull request by kurkomisi**
- `bundle.js` and `dashboard.html` is moved to the root directory.
- An own deep state updater replaces `ramda.js`.
    - The state of a `React` component is a JSON object, which needs to be immutable, and in order to keep this rule a state property needs to be changed completely during an update. This causes a problem, when one wants to nest objects in the state, because `React` doesn't change a small part of the nested object, but completely replaces it. The dashboard relies on such nested objects, which trigger the rerendering of some parts of the UI. In order to handle the deep update the `deepUpdate` function is used, which copies the original state property, and updates the concerned object properties. For the maximal flexibility of the update an auxiliary JSON object is used, which has the same structure as the state, except that it contains update handler functions.
   - It turns out that the server message and the state content can possess identical structure, so there is no need for handling the different parts of the incoming message, because the deep updater does it already. As a consequence an incoming message triggers all the relevant renders of the UI without further effort.
- `shouldUpdate` contains the structure of the incoming message, and it is used at the `shouldComponentUpdate` methods to check whether the data connected to the actual component was affected or not.
- Footer is included with the `geth` version on it.
- The message structure is changed both on server and client side.

### CODE REVIEW

**Comment by karalabe** [comment]
Ok, this is a bit crude on the UI side :P 

Instead of one big `1.8.0-unstable`, lets be a bit more specific with it. Something like:

```
Geth v1.8.0 (unstable)
Commit 12345678
```

Similar in styling to https://pbs.twimg.com/media/DMQdz7tW4AA_qlX.jpg:large

The `(unstable)` should only appear if unstable, and nothing should if it's a stable build. The commit hash is again optional, if we have one, display it, otherwise nothing. Also lets push it out to the right edge of the footer.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 53f04096a74238406109eacca62abc29684edbfb
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: dashboard/message.go [hunk 1]
```diff
@@ -29,7 +29,8 @@ type Message struct {
 }
 
 type GeneralMessage struct {
-	Version string `json:"version,omitempty"`
+	Version   string `json:"version,omitempty"`
+	GitCommit string `json:"gitCommit,omitempty"`
```
Comment (karalabe) [inline dashboard/message.go]: Call this just `Commit` pls, also in the json, just `commit`.

File: dashboard/dashboard.go [hunk 2]
```diff
@@ -69,7 +70,7 @@ type client struct {
 }
 
 // New creates a new dashboard instance with the given configuration.
-func New(config *Config) (*Dashboard, error) {
+func New(config *Config, gitCommit string) (*Dashboard, error) {
```
Comment (karalabe) [inline dashboard/dashboard.go]: `commit` is enough

File: dashboard/dashboard.go [hunk 3]
```diff
@@ -52,10 +52,11 @@ var nextID uint32 // Next connection id
...
 	config *Config
 
-	listener net.Listener
-	conns    map[uint32]*client // Currently live websocket connections
-	charts   *HomeMessage
-	lock     sync.RWMutex // Lock protecting the dashboard's internals
+	listener  net.Listener
+	conns     map[uint32]*client // Currently live websocket connections
+	charts    *HomeMessage
+	gitCommit string
```
Comment (karalabe) [inline dashboard/dashboard.go]: `commit` pls

File: dashboard/assets/types/content.jsx [hunk 4]
```diff
@@ -28,6 +28,7 @@ export type Content = {
 
 export type General = {
 	version: string,
+	gitCommit: string,
```
Comment (karalabe) [inline dashboard/assets/types/content.jsx]: `commit`

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: cbdb465003342451dbc7be5416c8e1146f7b878f
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
Please move commit under the version, style `Geth` and `Commit` a bit darker, make the fonts smaller, these aren't that important infos, just useful to have and move everything to the right side.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: cbdb465003342451dbc7be5416c8e1146f7b878f
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: cbdb465003342451dbc7be5416c8e1146f7b878f
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: cbdb465003342451dbc7be5416c8e1146f7b878f
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Pull request merged**